zebra-state/src/service.rs

```diff
@@ -1684,6 +1702,55 @@ impl Service<ReadRequest> for ReadStateService {
                 .map(|join_result| join_result.expect("panic in ReadRequest::ChainInfo"))
                 .boxed()
             }
+
+            #[cfg(feature = "getblocktemplate-rpcs")]
+            ReadRequest::CheckBlockProposalValidity(prepared) => {
+                let timer = CodeTimer::start();
+
+                let state = self.clone();
+
+                // # Performance
+                //
+                // Allow other async tasks to make progress while concurrently reading blocks from disk.
+                let span = Span::current();
+                tokio::task::spawn_blocking(move || {
+                    span.in_scope(move || {
+                        tracing::info!("attempting to validate and commit block proposal onto a cloned non-finalized state");
+                        let mut latest_non_finalized_state = state.latest_non_finalized_state();
+
+                        // Check that the previous block hash is the best chain tip if there are blocks in the 
+                        // non-finalized state. The previous block of a valid proposal must be on the best chain tip.
+                        if let Some(true) = latest_non_finalized_state.best_tip().map(|(_, tip_hash)| tip_hash != prepared.hash) {
+                            return Err("previous block hash is not the best chain tip".into())
+                        }
```

Comment:
I can see two potential issues here:
* we need to check the previous block hash in the proposal header, not the hash of the proposal header data
* it's a rare edge-case, but `validate_and_commit_non_finalized()` can also fork a non-finalized chain off the finalized tip

So I think what we might want here is something like:
```suggestion
                       // The previous block of a valid proposal must be on the best chain tip.
                       let Some((_best_tip_height, best_tip_hash) = read::best_tip(&latest_non_finalized_state, &state.db) {
                           if tip_hash != prepared.block.header.previous_block_hash {
                               return Err("proposal is not based on the current best chain tip: previous block hash must be the best chain tip".into());
                           }
                       } else {
                           return Err("state is empty: wait for Zebra to sync before submitting a proposal".into());
                       }
```

Reply:
> we need to check the previous block hash in the proposal header, not the hash of the proposal header data

🤦 Thank you.

> it's a rare edge-case, but validate_and_commit_non_finalized() can also fork a non-finalized chain off the finalized tip

I was thinking when the non-finalized state is empty, either a call to `commit_block` will fail to find the previous block hash and return `ValidateContextError::NotReadyToBeCommitted`, or `commit_new_chain` will commit the best chain to the non-finalized state.

> We should be able to validate directly off the finalized tip, if we want to. All the functions that we call below support it.

I'm still not sure how the non-finalized state could be empty near the network tip but if calling `validate_and_commit_non_finalized` when there is no best chain in the non-finalized state works I think we should keep it.

Reply:
> > it's a rare edge-case, but validate_and_commit_non_finalized() can also fork a non-finalized chain off the finalized tip
> 
> I was thinking when the non-finalized state is empty, either a call to `commit_block` will fail to find the previous block hash and return `ValidateContextError::NotReadyToBeCommitted`, or `commit_new_chain` will commit the best chain to the non-finalized state.
> 
> > We should be able to validate directly off the finalized tip, if we want to. All the functions that we call below support it.
> 
> I'm still not sure how the non-finalized state could be empty near the network tip but if calling `validate_and_commit_non_finalized` when there is no best chain in the non-finalized state works I think we should keep it.

I think it's a rare edge-case, it would only happen after a large chain fork, or if the node was disconnected from the network.